common/range-map/src/range_map.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::cmp::Ordering;
+use core::fmt::Debug;
+use core::fmt::{self};
+use core::ops::Bound;
+use core::ops::Range;
+use std::collections::BTreeMap;
+
+use super::range_wrapper::RangeWrapper;
+
+#[derive(Clone)]
+pub struct RangeMap<K, V> {
+    // Wrap ranges so that they are `Ord`.
+    // See `range_wrapper.rs` for explanation.
+    pub(crate) btm: BTreeMap<RangeWrapper<K>, V>,
+}
+
+impl<K, V> Default for RangeMap<K, V>
+where
+    K: Ord + Clone + std::fmt::Debug + Copy,
+    V: Eq + Clone + std::fmt::Debug,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub struct Iter<'a, K, V> {
+    inner: std::collections::btree_map::Iter<'a, RangeWrapper<K>, V>,
+}
+
+impl<'a, K, V> Iterator for Iter<'a, K, V>
+where
+    K: 'a,
+    V: 'a,
+{
+    type Item = (&'a Range<K>, &'a V);
+
+    fn next(&mut self) -> Option<(&'a Range<K>, &'a V)> {
+        self.inner.next().map(|(by_start, v)| (&by_start.range, v))
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.inner.size_hint()
+    }
+}
+
+impl<K: Debug, V: Debug> Debug for RangeMap<K, V>
+where
+    K: Ord + Clone + std::fmt::Debug + Copy,
+    V: Eq + Clone + std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_map().entries(self.iter()).finish()
+    }
+}
+
+impl<K, V> RangeMap<K, V>
+where
+    K: Ord + Clone + std::fmt::Debug + Copy,
+    V: Eq + Clone + std::fmt::Debug,
+{
+    /// Makes a new empty `RangeMap`.
+    pub fn new() -> Self {
+        RangeMap {
+            btm: BTreeMap::new(),
+        }
+    }
+
+    pub fn contain_key(&self, key: &K) -> Vec<(&RangeWrapper<K>, &V)> {
```

Comment:
Normally `contain_key` returns a bool. If it returns some stored value, it should be `get_xxx()`.